### PR TITLE
Not so small db and auth refactor

### PR DESCRIPTION
**Pull request by runtilov**
Changed quite a big chunk of the db structure, and such, the models, migrations, seeds, repositories and  even some routes.
Had checked to make sure that all work at least with GET, the rest I haven't had time to test out so there may be things that I broke.
As for the changes: 
  1. More info for objects returned from the backend
  2. Added a 'description' field for 'courses' and 'specialties'
  3. Replaced the old 'courses_teachers' tables with the 'lab_tasks' table renamed as 'course_instances' with an added field 'labTasksFileId' that is still not returned, but a boolean 'fileExists', indicating if the new field is null or not, is returned
  4. Added  a route for the table from point 3, '/course_instances' which supposedly takes the responsibility over the 'course_instances' table (no more access to this table from '/courses' and '/teachers', the new route should take care of everything
  5. Only the admin can create and delete rows in 'course_instances' thus binding a teacher and a course, the teacher can only update the file (maybe delete it as well, I forgot), the admin for now cannot change the file, but this will most likely change
  6. Added a M:M table between 'course_instances' and 'specialties' (instead of the planned one between 'groups' and 'teachers', seems like a better approach), named 'course_instances-specialties and added a route with a repository for it as well (no more repositories manipulating more than 1 table)
  7. Rerouted everything in consideration with the new tables, thus getting rid of the problem of one student having one or more teachers for the same course (I hope)
  8. Changed my formatter to 'prettier' which seems to be able to handle some edge case that my previous formatter couldn't, but due to this quite a few of the files changed a from me walking through files with my formatter

### CODE REVIEW

File: src/helpers/util.js [hunk 1]
```diff
@@ -59,110 +77,170 @@ function handleId(queryParams, Model, filter, queryParamsBindings) {
...
+      return dependentRow[`${parentName}Id`] === parentRow.id;
     });
-    model[`${dependentName}Count`] = matchingDependencies.length;
+    parentRow[`${dependentName}Count`] = matchingDependencies.length;
+  });
+
+  return rows;
+}
+
+async function appendDependentData(rows, /* models */{ parent, dependent }) {
```
Comment (alxolr) [inline src/helpers/util.js]: Very good naming, nice job!

File: src/helpers/util.js [hunk 2]
```diff
@@ -59,110 +77,170 @@ function handleId(queryParams, Model, filter, queryParamsBindings) {
...
+    for (let i = 0; i < currentRows.length; i += 1) {
+      const requiredParentId = currentRows[i][`${parentName}Id`];
+      const matchingParent = parents.find(parent => parent.id === requiredParentId);
+      if (!matchingParent) {
+        throw new errors.GeneralError('PARENT_NOT_FOUND', {
+          function: 'appendParentDataDeep',
+          file: './src/helpers/util.js',
+          cause: 'unknown',
+          severity: 'HUMAN_LOGIC_ERROR',
+          solution: 'tell Roman to fix it',
```
Comment (alxolr) [inline src/helpers/util.js]: 👍 Good one! I always like some humor in the code!

File: src/repositories/lab_reports.js [hunk 3]
```diff
@@ -19,38 +19,58 @@ module.exports = function createLabReportsRepository(mongoConnection, sqlConnect
...
-    labTaskId: [LabTask],
+    courseInstanceId: [CourseInstance],
     studentId: [Student],
-    labCommentId: [models.LabComment],
-    teacherId: [LabTask, models.Teacher],
-    courseId: [LabTask, models.Course],
+    teacherId: [CourseInstance, models.Teacher],
+    courseId: [CourseInstance, models.Course],
     groupId: [Student, models.Group],
+    // TODO: Is this the right route, there is a shorter one, but does that mean this is not the right one?
```
Comment (alxolr) [inline src/repositories/lab_reports.js]: Master Yoda yopta ?

**Comment by runtilov** [comment]
Câte puțin prin toate.
Era o problemă cu structura bazei de date așa că asta o cerut vreo câteva fișiere pentru tabele noi, plus modificări prin toate repositoriile pentru rutarea corectă a cererilor prin noua structură.
Nu-mi plăcea cum tabelele de legătură M:M erau manipulate din repositoriile responsabile de tabelele de pe lângă, așa că am pus repositorii și rute și pentru ele, astfel mutând responsabilități din vreo câteva repositorii vechi în acestea.
Am mai adăugat niște validări și pe lângă asta am pus un formatter nou și am umblat prin niște fișiere cu el, ce a mai adăugat nr de fișiere modificate.
E probabil să fi introdus bug-uri, fiindcă am mers cam rapid dat fiind că se apropie demo-ul general,
dar azi voi testa rutele să mă asigur că nu am încurcat nimic.

**Pull request merged**